integrated_channels/degreed/models.py

```diff
@@ -0,0 +1,273 @@
+        """
+        return DegreedLearnerExporter(user, self)
+
+    def get_course_data_transmitter(self):
+        """
+        Return a ``DegreedCourseTransmitter`` instance.
+        """
+        return DegreedCourseTransmitter(self)
+
+    def get_course_data_exporter(self, user):
+        """
+        Return a ``DegreedCourseExporter`` instance.
+        """
+        return DegreedCourseExporter(user, self)
+
+
+@python_2_unicode_compatible
+class DegreedLearnerDataTransmissionAudit(models.Model):
+    """
+    The payload we sent to Degreed at a given point in time for an enterprise course enrollment.
+    """
+
+    degreed_user_id = models.CharField(
+        max_length=255,
+        blank=False,
+        null=False
+    )
+
+    enterprise_course_enrollment_id = models.PositiveIntegerField(
+        blank=False,
+        null=False
+    )
+
+    course_id = models.CharField(
+        max_length=255,
+        blank=False,
+        null=False,
+        help_text="The course run's key which is used to uniquely identify the course for Degreed."
+    )
+
+    course_completed = models.BooleanField(
+        default=True,
+        help_text="The learner's course completion status transmitted to Degreed."
+    )
+
+    completed_timestamp = models.CharField(
+        max_length=10,
+        help_text=(
+            'Represents the Degreed representation of a timestamp: yyyy-mm-dd, '
+            'which is always 10 characters.'
+        )
+    )
+
+    # Request-related information.
+    status = models.CharField(max_length=100, blank=False, null=False)
+    error_message = models.TextField(blank=True)
+    created = models.DateTimeField(auto_now_add=True)
+
+    class Meta:
+        app_label = 'degreed'
+
+    def __str__(self):
+        """
+        Return a human-readable string representation of the object.
+        """
+        return (
+            '<DegreedLearnerDataTransmissionAudit {transmission_id} for enterprise enrollment {enrollment}, '
+            'Degreed user {degreed_user_id}, and course {course_id}>'.format(
+                transmission_id=self.id,
+                enrollment=self.enterprise_course_enrollment_id,
+                degreed_user_id=self.degreed_user_id,
+                course_id=self.course_id
+            )
+        )
+
+    def __repr__(self):
+        """
+        Return uniquely identifying string representation.
+        """
+        return self.__str__()
+
+    def serialize(self):
+        """
+        Return a JSON-serialized representation.
+
+        Sort the keys so the result is consistent and testable.
+        """
+        return json.dumps(self._payload_data(), sort_keys=True)
+
+    def _payload_data(self):
+        """
+        Convert the audit record's fields into Degreed key/value pairs.
+        """
+        return {
+            'employeeId': self.degreed_user_id,
+            'id': self.course_id,
+            'completionDate': self.completed_timestamp,
+        }
```

Comment:
This is another carry-over from the SAPSF version.

The issue is that, at least for course metadata exports, it's happening through the exporter itself, with all those `transform_x` functions -- I *really* like that way of doing things, as it makes way more sense from a higher level, and is easier to follow in the code.

However, for some strange reason, the learner data exporter was (by "was" I mean this is how it was all wired up before and still is, despite the refactorizations) 'exporting' data through use of the learner data audit model. I tried hard to find a way to untangle this and make things consistent with the other cleaner technique of serialization, but I quickly realized that would mean a *lot* of extra refactorizations, this time not to just move things around, but also to change a bunch of logic which may have bad side-effects and unforeseen bugs in the existing SuccessFactors code. Complete coverage wouldn't guarantee us success here, no pun intended.

So I think we can still do this, and I would prefer it, but it would delay the delivery by quite a bit, and increase testing complexity. The tradeoff for cleaning up this part of the architecture is just more time...

Reply:
OK, thanks for explaining the reasoning behind the current implementation. I'd be fine with doing additional refactoring as part of another ticket -- to make sure that this happens, though, maybe add `TODO`s to the docstrings of these methods, and create a follow-up ticket *now*?

CC @douglashall -- final decision up to you, of course :)

Reply:
@itsjeyd @douglashall A separate ticket would be the most reasonable given how big this PR already is, yeah. It's not going to affect functionality, but it's more of a continuation of the refactor to match the ideal, which we'd ultimately prefer at some point.

I will add some TODOs, and wait for @douglashall 's response to make the ticket.

Reply:
I agree, let's create a ticket describing the additional refactor and move on with what we have.

Reply:
@douglashall Great, I'll make a ticket soon!

Reply:
@douglashall @itsjeyd https://openedx.atlassian.net/browse/ENT-761 is made.

Reply:
P.S. at least for this case, I was able to make the serialize method return all the data needed now.